src/payment/data/sagas.js

```diff
@@ -19,10 +19,38 @@ import { handleErrors, handleMessages } from '../../feedback';
 export function* handleFetchBasket() {
   try {
     yield put(fetchBasket.request());
-    const result = yield call(PaymentApiService.getBasket);
+    let result = yield call(PaymentApiService.getBasket);
     yield put(fetchBasket.success(result));
     yield put(basketDataReceived(result));
     yield call(handleMessages, result.messages, true);
+    if (result.products.length === 1) {
+      const [product] = result.products;
+      const { courseKey } = product;
+      const discount = yield call(
```

Comment:
I'm wondering if we can off-load this work to the API itself so that the first time we call getBasket we just get all the information we need.  i.e., the original call to getBasket on line 22 would just include whether or not a first purchase discount is available.

Failing that, it might be cleaner to create a service method that encapsulates this logic/conditional calling of checkBasketForDiscount so that we can keep it out of the saga.  That would mean that on line 22 we'd come back here with all the data we need, and wouldn't have to worry about DRYing up all the logic here/re-dispatching all the redux actions and all that.

Reply:
for point 1, i believe @robrap has more context on this, but i believe the decision was not to have the data available on the first request, so that we wouldn't need to make a synchronous request to the lms from ecommerce

Reply:
for point 2, i think your suggestion is to get all the changes we need on line 22 before we complete rendering the basket the first time.  
i believe this would then essentially be a synchronous request, which i think @robrap said we can't do? again, i don't have the full context on this, but i think @robrap does
(and @emma-green)

Reply:
Apologies, this is hard to explain in text. :)

For point 2, I'm suggesting that the logic of 1) calling `getBasket`, 2) then `checkBasketForDiscount`, 3) then possibly `getBasket` again, could be extracted from this saga and hidden behind a single service method call on line 22, something like:

```
let result = yield call(PaymentApiService.getBasketWithPossibleDiscount) // terrible name`
```

And then the saga here would effectively stay the same as it originally was, without any knowledge that those extra calls were being made. Then in your new service method you could write the logic without all the extra syntax of this saga:

```
async function getBasketWithPossibleDiscount() {
  let result = await getBasket();
  if (result.products.length === 1) {
    const [product] = result.products;
    const { courseKey } = product;
    const discount = await checkBasketForDiscount(courseKey);
    if (discount.discount_applicable) {
      result = await getBasket(discount.jwt);
    }
    return result;
}
```

Let me know if that doesn't make sense - happy to try to explain with spoken words!

Reply:
thanks! i tried to implement what you had in slack.
omitting this after the second getBasket call seems to work, but what does this statement do?
yield put(fetchBasket.success(result));